File: apps/beavernest-be/tests/unit/Steps/ReadinessSteps.fs
module BeaverNestBe.Tests.Unit.Steps.ReadinessSteps

open System.Net.Http
open TickSpec
open Xunit
open BeaverNestBe.Application.ReadinessPort
open BeaverNestBe.WebApp
open BeaverNestBe.Tests.Unit.Steps.BddState

let mutable private client: HttpClient option = None
let mutable private lastStatus = 0
let mutable private lastBody = ""
let mutable private cacheControl = ""
let mutable private containsValidator = false

let private runReadiness port =
    let response =
        buildClient (webAppWith port)
        |> fun http -> http.GetAsync("/api/v1/readiness").Result

    lastStatus <- int response.StatusCode
    lastBody <- response.Content.ReadAsStringAsync().Result
    cacheControl <- response.Headers.GetValues("Cache-Control") |> Seq.head

    containsValidator <-
        response.Headers.Contains("ETag")
        || response.Content.Headers.Contains("Last-Modified")

[<Given>]
let ``startup migrations completed and SQLite accepts queries`` () = runReadiness alwaysReady

[<Given>]
let ``SQLite cannot complete the readiness query`` () =
    runReadiness (fromProbe (fun () -> false))

[<When>]
let ``I send a GET request to "/api/v1/readiness"`` () = ()

[<Then>]
let ``the response status is 200`` () = Assert.Equal(200, lastStatus)

[<Then>]
let ``the response status is 503`` () = Assert.Equal(503, lastStatus)

[<Then>]
let ``the JSON response reports status "ready", database "ready" and schema "current"`` () =
    Assert.Equal("{\"status\":\"ready\",\"components\":{\"database\":\"ready\",\"schema\":\"current\"}}", lastBody)

[<Then>]
let ``the JSON response reports status "not-ready"`` () =
    Assert.Equal(
        "{\"status\":\"not-ready\",\"components\":{\"database\":\"unavailable\",\"schema\":\"unknown\"}}",
        lastBody
    )

[<Then>]
let ``the response reveals no database path, SQL text or exception detail`` () =
    Assert.DoesNotContain("/", lastBody)
    Assert.DoesNotContain("select", lastBody, System.StringComparison.OrdinalIgnoreCase)
    Assert.DoesNotContain("exception", lastBody, System.StringComparison.OrdinalIgnoreCase)

[<Then>]
let ``the response sends "Cache-Control: no-store" without a cache validator`` () =
    Assert.Equal("no-store", cacheControl)
    Assert.False(containsValidator)

File: apps/beavernest-be-e2e/steps/readiness.steps.ts
/**
 * Aggregate HTTP observations for liveness and readiness. Detailed database
 * fault injection remains in the F# integration suite; this BDD surface
 * verifies the externally observable contract of the disposable runtime.
 */
import { expect } from "@playwright/test";
import { createBdd } from "playwright-bdd";
import { backendShell, requireComposeRuntime, resetBackendData } from "../utils/compose-runtime";
import { getResponse } from "../utils/response-store";
import { expectNoStorageDiagnostics, expectReadinessResponse } from "../utils/readiness";

const { Given, Then } = createBdd();

Given("the BeaverNest process is accepting HTTP requests", async () => {
  // The disposable runtime is started by scripts/run-e2e.sh before Playwright runs.
});

Given("startup migrations completed and SQLite accepts queries", async () => {
  // Startup is observed through the ready response below.
});

Given("SQLite cannot complete the readiness query", async () => {
  requireComposeRuntime();
  // Replacing only the disposable named-volume database forces the next
  // read-only readiness probe to fail, without adding a production test seam.
  await backendShell(
    "printf '%s' 'invalid sqlite fixture' > /var/lib/beavernest/beavernest.sqlite3.next && mv /var/lib/beavernest/beavernest.sqlite3.next /var/lib/beavernest/beavernest.sqlite3 && rm -f /var/lib/beavernest/beavernest.sqlite3-wal /var/lib/beavernest/beavernest.sqlite3-shm",
  );
});

// oxlint-disable-next-line no-empty-pattern
Then("the JSON response reports status {string}", async ({}, status: string) => {
  const body = (await getResponse().json()) as { status?: unknown };
  expect(body.status).toBe(status);
});

// oxlint-disable-next-line no-empty-pattern
Then(
  "the JSON response reports status {string}, database {string} and schema {string}",
  // oxlint-disable-next-line no-empty-pattern
  async ({}, status: string, database: string, schema: string) => {
    await expectReadinessResponse(getResponse(), status, database, schema);
  },
);

// oxlint-disable-next-line no-empty-pattern
Then("the response sends {string} without a cache validator", async ({}, cacheControl: string) => {
  const response = getResponse();
  const separator = cacheControl.indexOf(":");
  expect(separator).toBeGreaterThan(0);
  const headerName = cacheControl.slice(0, separator).toLowerCase();
  const expectedValue = cacheControl.slice(separator + 1).trim();
  expect(headerName).toBe("cache-control");
  expect(expectedValue).toBe("no-store");
  expect(response.headers()[headerName.toLowerCase()]).toBe(expectedValue);
  expect(response.headers().etag).toBeUndefined();
  expect(response.headers()["last-modified"]).toBeUndefined();

  if (response.status() === 503) {
    await resetBackendData();
  }
});

Then("the response reveals no database path or migration detail", async () => {
  await expectNoStorageDiagnostics(getResponse());
});

Then("the response reveals no database path, SQL text or exception detail", async () => {
  await expectNoStorageDiagnostics(getResponse());
});
